File: aoc-py-2021/day13.py
import re
data = open('day13.txt').read()
coords = [(int(x),int(y)) for x,y in re.findall('(\d+),(\d+)',data)]
splits = [(ax,int(amt))for ax,amt in re.findall('([xy])=(\d+)',data)]

def fold_coord(coord,axis,split_ix):
  x,y = coord
  if axis == 'x':
    return (x - 2 * (x - split_ix), y) if x - split_ix > 0 else (x,y)    
  return (x, y - 2 * (y - split_ix)) if y - split_ix > 0 else (x,y)

for i,(axis,split_ix) in enumerate(splits):
  coords = set(fold_coord(c,axis,split_ix) for c in coords)
  if i == 0:print(len(set(coords))) # part 1

def display(coords):
  maxx,maxy = max(x for x,y in coords), max(y for x,y in coords)
  empty = [[' ' for col in range(maxx+1)] for rows in range(maxy+1)]
  for x,y in coords:
    empty[y][x] = '#'
  for p in empty:print("".join(p))

display(coords)

File: aoc-clj-2021/src/aoc_clj_2021/day13.clj
(ns aoc-clj-2021.day13
 (:require  [clojure.string :as str]
            [clojure.set :as set]
            [clojure.java.io :as io]))

(def data (slurp (io/resource "day13.txt")))

(def coords
  (->> (re-seq #"(\d+),(\d+)" data)
       (map rest)
       (map #(map parse-long %))
       (set)))

(def folds
  (->> (re-seq #"([xy])=(\d+)" data)
       (map rest)
       (map (fn [[axis n]] [axis (parse-long n)]))))

(defn fold-coord [[x y] axis split-ix]
  (let [xadj (- x (* 2 (- x split-ix)))
        yadj (- y (* 2 (- y split-ix)))]
    (cond (and (= axis "x") (> x split-ix)) [xadj,y]
          (and (= axis "y") (> y split-ix)) [x,yadj]
          :else [x,y])))

(defn display-code [coords]
  (let [maxx (apply max (map first coords))
        maxy (apply max (map second coords))
        disp (vec (repeat (inc maxy) (vec (repeat (inc maxx) " "))))
        full (reduce #(assoc-in %1 (reverse %2) "#") disp coords)]
      (doseq [ln full] (println (str/join ln)))))

(let [[part1 part2-coords]
    (->> (reductions (fn [c [ax n]] (set (map #(fold-coord % ax n ) c)))
                    coords
                    folds)
        ((juxt (comp count second) last)))]
   (println part1)
   (display-code part2-coords))
